client/network/src/persist_peers.rs

```diff
@@ -0,0 +1,308 @@
+						})
+						.collect::<Vec<_>>();
+					(protocol.to_owned(), entries)
+				})
+				.collect();
+
+			let busy_future = persist_peer_addrs::persist(Arc::clone(&self.paths), entries).boxed();
+			self.busy = Some(busy_future);
+		}
+		Poll::Pending
+	}
+}
+
+mod persist_peer_addrs {
+	use super::*;
+	use tokio::io::AsyncWriteExt;
+
+	#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+	pub(super) struct PeerEntry {
+		pub peer_id: String,
+		pub addrs: Vec<Multiaddr>,
+	}
+
+	pub(super) async fn persist(
+		paths: Arc<Paths>,
+		protocols: HashMap<ProtocolType, Vec<PeerEntry>>,
+	) -> Result<(), io::Error> {
+		let mut tmp_file = tokio::fs::OpenOptions::new()
+			.create(true)
+			.write(true)
+			.open(&paths.tmp_path)
+			.await?;
+		let serialized = serde_json::to_vec_pretty(&protocols)?;
+
+		tmp_file.write_all(&serialized).await?;
+		tmp_file.flush().await?;
+		std::mem::drop(tmp_file);
+
+		tokio::fs::rename(&paths.tmp_path, &paths.path).await?;
+
+		Ok(())
+	}
+
+	pub(super) fn load(
+		path: impl AsRef<Path>,
+	) -> Result<HashMap<ProtocolType, Vec<PeerEntry>>, io::Error> {
+		let file = match std::fs::OpenOptions::new().read(true).open(path.as_ref()) {
+			Ok(file) => file,
+			Err(not_found) if not_found.kind() == std::io::ErrorKind::NotFound =>
+				return Ok(Default::default()),
+			Err(reason) => return Err(reason),
+		};
+		let entries = serde_json::from_reader(file)?;
+		Ok(entries)
+	}
+}
+
+pub struct PersistPeersets(BoxedFuture<Never>);
+pub use peersets::load as peersets_load;
+
+impl PersistPeersets {
+	pub fn new(dir: impl AsRef<Path>, peerset_handle: PeersetHandle) -> Self {
+		let paths = Paths::new(dir, "peer-sets");
+		let busy_future = async move {
+			let mut ticks = tokio::time::interval(FLUSH_INTERVAL);
+			ticks.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Skip);
+
+			loop {
+				let _ = ticks.tick().await;
+				if let Err(reason) = peersets::persist(&paths, &peerset_handle).await {
+					log::warn!("Error persisting peer sets: {}", reason);
+				}
+			}
+		};
+		Self(busy_future.boxed())
+	}
+
+	pub fn poll(&mut self, cx: &mut Context) -> Poll<Never> {
+		self.0.poll_unpin(cx)
+	}
+}
+
+mod peersets {
+	use super::*;
+
+	#[derive(Debug, serde::Serialize, serde::Deserialize)]
+	pub struct PeerInfo {
+		pub peer_id: String,
+		pub reputation: i32,
+		pub sets: Vec<usize>,
+	}
+
+	pub(super) async fn persist(
+		paths: &Paths,
+		peerset_handle: &PeersetHandle,
+	) -> Result<(), io::Error> {
+		use tokio::io::AsyncWriteExt;
+
+		let peersets_dumped = peerset_handle
+			.dump_state()
+			.await
+			.map_err(|()| io::Error::new(io::ErrorKind::BrokenPipe, "oneshot channel failure"))?
+			.into_iter()
+			.map(|(peer_id, reputation, sets)| PeerInfo {
+				peer_id: peer_id.to_base58(),
+				reputation,
+				sets,
+			})
+			.collect::<Vec<_>>();
+
+		let mut tmp_file = tokio::fs::OpenOptions::new()
+			.create(true)
+			.write(true)
+			.open(&paths.tmp_path)
```

Comment:
This is missing a `.truncate(true)`. If there's a stale temporary file here this will only overwrite the beginning of the file, resulting in a corrupted JSON.

See [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=310279586e0e296201f75606ce13e126) for an example of how this works.